### PR TITLE
KAA-1129: 'Endpoint groups' page

### PR DESCRIPTION
**Pull request by v-chyzhevskyi**
'Endpoint groups' page

### CODE REVIEW

**Comment by ashvayka** [comment]
use
- TOC
  {:toc}
  to generate TOC instead of manual work

**Comment by ashvayka** [comment]
"Profile Filters" section should be above "Using endpoint groups". Otherwise it is very hard to read. You use the same schema in the start and in the end of the page. In the middle of the page you use different schema

**Comment by ashvayka** [comment]
Several major comments:
1) Add TODO for all REST API links (concrete links need to be provided when @sashadidukh and @DimaMir999 will finish automation of REST API generation)
2) Validate all Admin UI links.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 1]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+|iOS devices with premium subscription plan | #cp.os.toString().equals("iOS") and #sp.subscriptionPlan.toString().equals("Premium")         |false	 |false |true  |
+
+## Adding endpoint groups ##
+
+Endpoint groups are created based on the profile filter.
+To add a new endpoint group, do the following:
+
+1. Open the **Endpoint groups** window by clicking **Endpoint groups** under the application menu on the navigation panel and then click **Add endpoint group**.
+
+<img src="admin-ui/endpoint-groups.png">
```
Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: Don't use html tags for image. Use markdown syntax for that.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 2]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
-Endpoint group profile filters are predicate expressions which define characteristics of group members (endpoints). These filters are executed against the endpoint profile to figure out whether or not the endpoint belongs to the group.
+* TOC
+{:toc}
 
-### NOTE ###
+Kaa provides a mechanism for endpoints aggregation within the application that is based on groups. Grouping endpoints enables you to activate specific configuration parameters, 
+control access to notification topics, etc.
+This guide will familiarize you with the basic concepts of designing endpoint groups. It is assumed that you have either set up a Kaa Sandbox, or a fully-blown Kaa cluster 
+already and that you have created a tenant and an application in Kaa. It is also strongly recommended to review
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: that you have created a tenant and an application in Kaa.

Can here be added a link to the part that describes how to achieve this requirement?

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 3]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
-### NOTE ###
+Kaa provides a mechanism for endpoints aggregation within the application that is based on groups. Grouping endpoints enables you to activate specific configuration parameters, 
+control access to notification topics, etc.
+This guide will familiarize you with the basic concepts of designing endpoint groups. It is assumed that you have either set up a Kaa Sandbox, or a fully-blown Kaa cluster 
+already and that you have created a tenant and an application in Kaa. It is also strongly recommended to review 
+[Endpoint profiles]({{root_url}}Programming-guide/Key-platform-features/Endpoint-profiles/) before you proceed with this one. 
 
-	Different profile schema versions may require separate profile filters due to the schema structural differences. 
-	In case a group has no filter assigned for a specific profile schema version, the group does not apply to the endpoints that use the profile of this schema version.
+Endpoint profile is a virtual identity or "passport" of the endpoint. By filtering against the data in the profiles, endpoints may be aggregated into independently managed groups.
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: It looks a bit strange to see recommendation for visiting of the Endpoint profiles documentation page and then a short description if this entity.

Could this be re-formatted to something like below?
1) Replace "It is also strongly recommended ... before you proceed with this one." with "It is strongly recommended to get familiar with Endpoint profiles before you proceed." and drop the link there.
2) Add the link to the Endpoint profiles documentation page at the beginning of the next paragraph (like "[Endpoint profiles]({{root_url}}Programming-guide/Key-platform-features/Endpoint-profiles/) is a virtual identity or "passport" of the endpoint."

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 4]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+Kaa allows for aggregating endpoints related to the same application into endpoint groups. The _endpoint group_ represents an independent managed entity which is
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: The 3 sentences:
- By filtering against the data in the profiles, endpoints may be aggregated into independently managed groups.
- Kaa allows for aggregating endpoints related to the same application into endpoint groups.
- The _endpoint group_ represents an independent managed entity which is defined by the profile filters assigned to the group.

Look unrelated and bring misunderstanding. Can this be re-written for clarity?

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 5]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+defined by the profile filters assigned to the group. Those endpoints whose profiles match the profile filters of the specific endpoint group become automatically 
+registered as members of this group. There is no restriction for endpoints on having membership in a number of groups at a time.
 
-Every endpoint group within the application has a unique weight value assigned to it. The weight value is used to resolve conflicts among endpoint groups: in general, the larger the weight, the higher the group priority.
-Every Kaa application comes with the default, non-user-editable group "all", with the weight value 0. The profile filter of this group is automatically set to "true" for every profile schema version in the system. As a result, the "all" group contains every endpoint registered in the application. The "all" group is used to define the default configuration, default notification topics access list, and for some other special functions.
+Endpoint group profile filters are predicate expressions which define characteristics of group members (endpoints). These filters are executed against the endpoint 
+profile to figure out whether or not the endpoint belongs to the group
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: ... to figure out whether or not the endpoint belongs to the group**.**('dot' is missing)

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 6]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
 
+**NOTE**: Different profile schema versions may require separate profile filters due to the schema structural differences. 
+In case a group has no filter assigned for a specific profile schema version, the group does not apply to the endpoints that use the profile of this schema version.
 
 ## Profile filters ##
 
-
-Profile filters in Kaa are based on the [Spring Expression Language](http://docs.spring.io/spring/docs/3.0.x/reference/expressions.html) (SpEL). All filters must be specified as predicates (statements which may be either true or false). For further reference on the filters syntax, please refer to the Spring documentation.
+Profile filters in Kaa are based on the [Spring Expression Language](http://docs.spring.io/spring/docs/3.0.x/reference/expressions.html) (SpEL).
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:36]: Can the link to SpEL documentation be moved to the sentence that refers it while this one can to be a more generic and short description or definition of the Spring Expression Language only?

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 7]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+In case a group has no filter assigned for a specific profile schema version, the group does not apply to the endpoints that use the profile of this schema version.
 
 ## Profile filters ##
 
-
-Profile filters in Kaa are based on the [Spring Expression Language](http://docs.spring.io/spring/docs/3.0.x/reference/expressions.html) (SpEL). All filters must be specified as predicates (statements which may be either true or false). For further reference on the filters syntax, please refer to the Spring documentation.
+Profile filters in Kaa are based on the [Spring Expression Language](http://docs.spring.io/spring/docs/3.0.x/reference/expressions.html) (SpEL). 
+All filters must be specified as predicates (statements which may be either true or false). 
+For further reference on the filters syntax, please refer to the Spring documentation.
 Profile filters are evaluated using following context variables:
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:38]: Can this part be a bit more explained? At the least, a reference to the filter explanation below can be added.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 8]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
 * "cp" - Client-side endpoint profile
 * "sp" - Server-side endpoint profile
 * "ekh" - Endpoint key hash
 
-## Profile filter examples ##
+### Profile filter examples ###
 
 The following example illustrates the general idea of profile filters.
 
+1. Let's assume the following client-side profile schema.
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:50]: Could this be extended with the information within the schema? A reader will need at least minimal understanding of what the schema is and consists of.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 9]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+               "null"
+            ]
+         }
+      ]
+   }
+]
+```
 
+2. Let's assume the following server-side profile schema. Please note that this schema is less complex only for demonstration purposes. 
+Both Client-side and Server-side profile schemas support same level of complexity.
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: "Both Client-side and Server-side profile schemas support same level of complexity."

Is that a feature by itself? The profiles should be flexible and support virtually any configuration. The complexity is a side-effect of this and should not be a feature. 

Can this be re-written?

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 10]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+|#cp.recordField.otherMapSimpleField.size()==2 | The client-side endpoint profile contains recordField, which is a record containing otherMapSimpleField, which is a collection containing two entries. |
+|#cp.arrayRecordField[1].otherSimpleField==789 |The client-side endpoint profile contains arrayRecordField, which is an array. This array contains an element in the position 1, which is a record containing otherSimpleField set to '789'.|
+|#cp.nullableRecordField==null |An example of how to check a field for the null value.|
+|#cp.arraySimpleField[0]=='CLIENT_SIDE_VALUE_1' and # sp.arraySimpleField[0]=='SERVER_SIDE_VALUE_1'|An example of how to combine several conditions in a query.|
+|!#arrayRecordField.?[otherSimpleField==456].isEmpty() |The arrayRecordField field is an array of records. It contains at least one element that contains otherSimpleField with the value 
+
+## Using endpoint groups ##
+
+Each Kaa application has a special, built-in, non-user-editable group "all" with weight 0. The associated profile filter is automatically set equal to "true" for each profile
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: Could here be added some information of what other properties groups have, what is group weight, etc? Should this information be available below, please add a short description here and a reference to more details there.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 11]
```diff
@@ -6,166 +6,308 @@ nav: /:path/Programming-guide/Key-platform-features/Endpoint-groups
...
+
+4. Click **Save** to save the profile filter.
+   
+    **NOTE**: You can save the data on the **Draft** tab and return to update it later as many times as needed until you click **Activate**.
+   
+5. Click **Activate** to activate the profile filter.
+   
+   All the specified information will be displayed on the **Active** tab.
 
+## REST API ##
```
Comment (Slash32) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:324]: The page lacks Further reading block.

**Comment by Slash32** [comment]
The page seems incomplete and is composed in an unusual order which requires reader to re-read the page few times. I'd love to see the page re-composed appropriately.

**Comment by ashvayka** [comment]
It seems to me that page is structured logically. How do you propose to re-compose it?

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 12]
```diff
@@ -274,16 +286,17 @@ To add a new endpoint group, do the following:
...
 1. Open the **Endpoint groups** window by clicking **Endpoint groups** under the application menu on the navigation panel and then click **Add endpoint group**.
 
-<img src="admin-ui/endpoint-groups.png">
+![endpoint-groups](admin-ui/endpoint-groups.png "endpoint-groups")
+
 
 2. In the **Add endpoint group** window, fill in the required fields and then click **Add**.
 
-<img src="admin-ui/create-endpoint-group.png">
+![create-endpoint-group](admin-ui/create-endpoint-group.png "create-endpoint-group")
```
Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:294]: Don't add stupid alt text and title -- they are meant for people. Please, read https://en.wikipedia.org/wiki/Alt_attribute before continuing.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 13]
```diff
@@ -184,7 +186,17 @@ At last, the following filters will yield true when applied to the given endpoin
...
+Also group has:
+
+* Name
+* Weight
+* Description
+* [Profile filters](#profile-filters)
+* [Configurations]({{root_url}}Programming-guide/Key-platform-features/Configuration-management/#configuration-schema)
+* [Notification topics]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/#notification-topics)
+
+The associated profile filter is automatically set equal to "true" for each profile
```
Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md]: `true` should be enclosed in backticks:

```
`true`
```

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 14]
```diff
@@ -310,4 +323,12 @@ To add a profile filter to the endpoint group, do the following:
...
\ No newline at end of file
+Use [Admin REST API]({{root_url}}Programming-guide/Server-REST-APIs/#TODO) for getting more information.
+
+## Further reading ##
+
+* [Spring Expression Language](http://docs.spring.io/spring/docs/3.0.x/reference/expressions.html)
+
+* [Avro](http://avro.apache.org/)
+
+
```
Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:334]: Trailing empty lines.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md [hunk 15]
```diff
@@ -184,7 +186,17 @@ At last, the following filters will yield true when applied to the given endpoin
 
 ## Using endpoint groups ##
 
-Each Kaa application has a special, built-in, non-user-editable group "all" with weight 0. The associated profile filter is automatically set equal to "true" for each profile 
+Each Kaa application has a special, built-in, non-user-editable group "all" with weight 0. Weight of group responsible for her priority. Group with the biggest weight is the most priority.
```
Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:189]: s/her/its

Comment (rasendubi) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:189]: Check the grammar, please. There are multiple errors.

File: doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md (hunk 6 above)
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Endpoint-groups/index.md:36]: This is the first reference of SpEL on this page, right? There is a table with filter examples. I believe this is sufficient.

**Comment by Slash32** [comment]
Looks good to me

**Pull request merged**